elastic/logstash
# Plugin throughput, worker utilization and worker cost per event flow metrics.

## Release notes

[rn:skip]

## What does this PR do?
Followings are addressed:
- ~~Namespace store is shaped with RubySymbol key but filter and output codecs were using string key. This commit intends to standardize the namespace key with RubySymbol for filter & output codecs.~~
- Initializes throughput flow metrics for the input plugins.
- Initializes the worker cost per event and worker utilization for the filter and output plugins with only uptime metrics but it should combine with worker count, will be implemented in next commits.
- Fetching codec ID generated in ruby scope is possible but problematic to in Java scope. We will skip codec flow metrics since they are rarely produce the hard times.


## Why is it important/What is the impact to the user?
- See the https://github.com/elastic/logstash/issues/14463 meta proposal and sub task definition https://github.com/elastic/logstash/issues/14601 for the objectives.
- No impact to end users.

## Checklist




## Author's Checklist

Need to address to complete:

## How to test this PR locally

- Pull the changes locally, compile and run
- Call the stats APIs
- API response should include plugin throughput for the input plugins, worker cost & utilization for the filter & output plugins.

## Related issues

- Closes https://github.com/elastic/logstash/issues/14601

## Use cases



## Screenshots
<img width="629" alt="Screen Shot 2022-11-14 at 10 17 30 AM" src="https://user-images.githubusercontent.com/99575341/201581169-c33afbe3-b70a-452d-b747-ab1212bbb744.png">

<img width="584" alt="Screen Shot 2022-11-14 at 7 09 51 PM" src="https://user-images.githubusercontent.com/99575341/201681138-bb64f4e8-4781-4c16-8cc7-9a243874fb34.png">

- Generated Doc
<img width="876" alt="Screen Shot 2022-11-17 at 12 07 13 PM" src="https://user-images.githubusercontent.com/99575341/202379469-026e8c32-a299-4c55-8bb2-3f04b524d8d0.png">



## Logs



```
// integration test logs
 Starting Logstash /Users/mashhur/Dev/elastic/logstash/build/logstash-8.6.0-SNAPSHOT/bin/logstash -e input { stdin {} } filter { mutate { add_tag => 'integration test adding tag' } } output { stdout {} }
    Starting Logstash: /Users/mashhur/Dev/elastic/logstash/build/logstash-8.6.0-SNAPSHOT/bin/logstash ["-e", "input { stdin {} } filter { mutate { add_tag => 'integration test adding tag' } } output { stdout {} }"] (pwd: /Users/mashhur/Dev/elastic/logstash/qa/integration)
    Logstash started with PID 17020, LS_JAVA_HOME: /Users/mashhur/.sdkman/candidates/java/17.0.4-tem
      should retrieve plugin level flow metrics

```
